Unit tests

nunesjoab opened the pull request:
* **Please check if the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
This PR introduces a set of unit tests to iotagent-mosca


* **What is the current behavior?** (You can also link to an open issue here)



* **What is the new behavior (if this is a feature change)?**



* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)


* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)
dojot/dojot/1187

* **Other information**:

mprevide commented on coverage/clover.xml:
```diff
@@ -0,0 +1,408 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Files within coverage/ really need to be versioned?

nunesjoab replied:
I don't think so but I'm not sure if they are used by other tools like Travis or Codefactor. I will add the coverage path to gitignore. If necessary, I can remove this line later.

giovannicuriel replied:
After running the unit tests, codecov uploads the results to its site. There's no need track them in git.

mprevide commented:
I added codecov to my PR draft (https://github.com/dojot/iotagent-mosca/pull/74/commits/ac9aab23f682c4437bb1955602e2aa8af4c868ac), so maybe it's a good idea to add in this PR too. And I associated the project with the codecov service.

nunesjoab commented:
> I added codecov to my PR draft ([ac9aab2](https://github.com/dojot/iotagent-mosca/commit/ac9aab23f682c4437bb1955602e2aa8af4c868ac)), so maybe it's a good idea to add in this PR. And I associated the project with the codecov service.

Added :)

mprevide commented:
Maybe it's necessary remove coverage from gitignore and add something like below in package.json: 


"jest": {
    "coverageDirectory": "./coverage/",
    "collectCoverage": true,
    "collectCoverageFrom": [
      "lib/**/*.js"
    ]
  }

mprevide commented:
I think you should actually remove /covere/ * from gitignore.

giovannicuriel commented on src/iotagent.js:
```diff
@@ -128,7 +128,15 @@ class IoTAgent {
+      }
+    }
+
+    return undefined;
```
You can also just `return`.

jonaphael commented on tests/unit-tests/iotagent.test.js:
```diff
@@ -0,0 +1,237 @@
+
+        expect(iotagent._processDeviceRemoval('admin', event)).toBeUndefined();
+    });
+
```
to increase the add this case to test device actuation function

` // test a device actuation
 it("should process a device actuation ", () => {

        const tenant = 'my-tenant';

        const event = {
            data: {
                id: 1,
                attrs: {
                    'attr-1': 'attr-1-value'
                }
            }
        };

        const deviceID = event.data.id;
        const topic = `/${tenant}/${deviceID}/config`;

        // messages builded
        const message = {
            'topic': topic,
            'payload': JSON.stringify(event.data.attrs),
            'qos': 0,
            'retain': false
          };

          // >> Tested code
        const iotagent = new IoTAgent();
        iotagent.mqttBackend.server.publish = jest.fn();
        iotagent._processDeviceActuation(tenant, event);
        // << Tested code

        const firstArgument = iotagent.mqttBackend.server.publish.mock.calls[0][0];

        // >> Result verification
        expect(firstArgument).toEqual(message);
        // << Result verification
    })`

nunesjoab replied:
Added

jonaphael commented on tests/unit-tests/metrics.test.js:
```diff
@@ -0,0 +1,82 @@
+
+        const metricsAttribute = "connectedClients";
+        metrics.preparePayloadObject(metricsAttribute, 20);
+        expect(metrics.lastMetricsInfo[`${metricsAttribute}`]).toEqual("20");
```
this should be another test case because this case expects null.

nunesjoab replied:
Separated to another it

jonaphael commented on tests/unit-tests/metrics.test.js:
```diff
@@ -0,0 +1,82 @@
+        metricStore.lastMetricsInfo = null;
+        app.use(moscaMestrics.getHTTPRouter(metricStore));
+        response = await get('/iotagent-mqtt/metrics', '');
+        expect(response.status).toEqual(500);
```
separate 200 test case from 500

nunesjoab replied:
test of status 500 removed because stopApp can't be called to change value of isInitialized variable.

nunesjoab commented on tests/unit-tests/iotagent.test.js:
```diff
@@ -0,0 +1,237 @@
+    });
+
+    it("Should build a MqttBackend instance with support for unsecure connections", () => {
+        // const config = JSON.parse(JSON.stringify(defaultConfig));
```
Removed

nunesjoab commented on tests/unit-tests/iotagent.test.js:
```diff
@@ -0,0 +1,237 @@
+            payload: { toString: () => "sample-payload" }
+        }
+
+        expect(iotagent).toBeDefined();
```
removed

nunesjoab commented on tests/unit-tests/iotagent.test.js:
```diff
@@ -0,0 +1,237 @@
+        }
+
+        expect(iotagent).toBeDefined();
+        expect(iotagent.mqttBackend).toBeDefined();
```
Removed

nunesjoab commented on tests/unit-tests/iotagent.test.js:
```diff
@@ -0,0 +1,237 @@
+
+        iotagent.initHealthCheck(healthcheck.healthChecker);
+        expect(iotagent.messageMonitor.monitor).toBeDefined();
+        // expect(iotagent._registerCallbacks()).toBeTruthy();
```
Removed

nunesjoab commented on tests/unit-tests/iotagent.test.js:
```diff
@@ -0,0 +1,237 @@
+ * - mosca
+ */
+
+// var Agent = require('@dojot/iotagent-nodejs').IoTAgent;
```
Removed

nunesjoab commented on tests/unit-tests/iotagent.test.js:
```diff
@@ -0,0 +1,237 @@
+        expect(iotagent.mqttBackend).toBeDefined();
+        expect(iotagent.getTopicParameter(samplePacket.topic, 1)).toEqual('Yoda');
+        expect(iotagent.getTopicParameter()).toBeUndefined();
+        expect(iotagent.getTopicParameter('topic')).toEqual('topic');
```
Separated to another It
